### ISSUE TITLE
Unable to JSON serialize type with optional keys

### ISSUE THREAD

**Issue by parkan** [dbartholomae/lambda-middleware#43]
## Describe the bug
The json serializer works fine for object literals, however if passed a type with optional members (I generally return models from my handlers, which usually have optional fields), it doesn't typecheck.

## To Reproduce
```
  describe("with a handler returning a type with optional members", () => {
    let response: any;

    type Thing = {
        id?: string
        foo: string
    }

    beforeEach(async () => {
      const handlerResponse: Thing = { foo: "bar" };
      const handler = async () => handlerResponse;
      const handlerWithMiddleware = jsonSerializer()(handler);
      response = await handlerWithMiddleware(createEvent({}), createContext());
    });

    it("returns 200", async () => {
      expect(response).toMatchObject({ statusCode: 200});
    });

    it("returns serialized non-optional members", async () => {
      expect(response.body).toEqual("{\"foo\":\"bar\"}");
    });
  })
```

yields

```
    src/JsonSerializer.test.ts:38:54 - error TS2345: Argument of type '() => Promise<Thing>' is not assignable to parameter of type 'PromiseHandler<APIGatewayProxyEvent, { [key: string]: JSONPrimitive; } | JSONObject[] | undefined>'.
      Type 'Promise<Thing>' is not assignable to type 'Promise<{ [key: string]: JSONPrimitive; } | JSONObject[] | undefined>'.
        Type 'Thing' is not assignable to type '{ [key: string]: JSONPrimitive; } | JSONObject[] | undefined'.
          Type 'Thing' is not assignable to type '{ [key: string]: JSONPrimitive; }'.
            Property 'id' is incompatible with index signature.
              Type 'string | undefined' is not assignable to type 'JSONPrimitive'.
                Type 'undefined' is not assignable to type 'JSONPrimitive'.
```

## Expected behavior
Type check passes as the optional key is still of the right type.

## Additional context
This appears to be a bit of a typescript wart, as [optional members are treated as equivalent to having `| undefined` under strict null checks](https://github.com/microsoft/TypeScript/issues/13778), which doesn't really make sense to me given that even in underlying JS hasOwnProperty can distinguish the two cases.

This seems like it can be fixed by redefining `type JSONPrimitive = string | number | boolean | JSONObject | undefined;`, though this isn't technically valid JSON, JSON.stringify has a relatively defined behavior for it.

I *think* there is a way to do this cleanly with keyof/Pick, trying to work it out...

EDIT: I should also note that this doesn't work at all if the returned value is typed as `interface`; that seems to really break the index signature inference

**Comment by parkan** [comment]
ok, I am able to make this mostly work correctly, though the overall test suite fails on my system:

types/JSONObject.ts:
```
export type RequiredKeys<T> = { [k in keyof T]-?: undefined extends T[k] ? never : k }[keyof T];
```

JsonSerializer.ts:
```
import { JSONObject, RequiredKeys } from "./types/JSONObject";

const logger: IDebugger = debugFactory("@lambda-middleware/json-serializer");

export const jsonSerializer = () => (
  handler: PromiseHandler<APIGatewayEvent, RequiredKeys<JSONObject> | undefined>
...
```

**Comment by parkan** [comment]
BTW thanks for being so responsive on these libraries, it seems like you're the only maintainer so I really appreciate it as it as they really fit my usecase and I'm frankly surprised they are not more widely used. This isn't strictly as "bug" as much as a limitation, but there was no better template to use. I'm happy to PR with these changes if they make sense to you, I'm still relatively new to typescript (mostly Scala background for this kind of thing) so I am not entirely sure of myself and appreciate the guidance.

**Comment by dbartholomae** [comment]
Thanks! I've quickly fixed it myself, as the setup with Rush currently is a bit finicky.

**Comment by dbartholomae** [comment]
It's released in version 2.1.0
https://www.npmjs.com/package/@lambda-middleware/json-serializer
